### ISSUE TITLE
Feature Request: Move image when trained and update name

### ISSUE THREAD

**Issue by mr-manuel** [jakowenko/double-take#84]
Hello,

first of all, thank you for your amazing work!

I would like to propose this function: When in the config the setting `save.unknown` is `true` (default) and I train an unknown image in the matches section it would be great, if the image is automatically moved/copied to `.storage/train/{matchname}`.

It would also be nice if a "modified" (trained/untrained) image would be reevaluated again, so that the unknown images are not unknown again or if known images are corrected with the right name.

I'm using the latest beta as of now.

What do you think about this?

**Comment by jakowenko** [comment]
Hey @mr-manuel, thank you for checking out my project and for the feature suggestions. I think it makes sense and is a great idea so I took a stab at implementing the reprocessing functionality.

Any images you train from the matches page will be copied and used on the training page. This was already built in and should be working. I believe that is what you were describing for the first function, right?

Here's a little gif demonstrating how I'm handling the reprocessing of the images. Currently you have to do it on an image by image basis. There is a button in the bottom right of every match image. Clicking this will reprocess that image and update the results.

In this gif you can see the first image was for an unknown user. I train it for myself, then reprocess the image to show it has the updated results.

![2021-09-03 02 38 56](https://user-images.githubusercontent.com/1081811/131961960-8ac39e6b-4c3c-4bd6-ac4d-78712c6fc213.gif)

**Comment by mr-manuel** [comment]
Hi @jakowenko, thank you for your fast answer!

First function: Yes that was exactly that what I was describing. But somehow this don't work. The images don't get copied after training in the matches section to the train folder. In the logs I can't find any error. Probably something with permissions, even if Double Take managed to create the folders? For testing purpose i also changed the folder permissions to 777 and file permissions to 666, but no change.

I also copied manually the files to the train folders. Double Take recognized them only after a restart of the container (could be normal since I did not use the upload function), but afterwards they disappeared under matches, even if the files are still in the matches folder.

```
# ls -l matches/
total 1232
-rw-r--r-- 1 root root 97861 Sep  2 17:07 007ba212-f92d-4fe4-ba27-b1b3e4bcf230.jpg
-rw-r--r-- 1 root root 75459 Sep  2 19:00 0dc61d88-f05e-4b04-8486-eedd6c3e1d95.jpg
-rw-r--r-- 1 root root 61028 Sep  3 08:58 17a0f83b-9799-429a-9e81-a19bfe2feef1.jpg
-rw-r--r-- 1 root root 62722 Sep  3 09:04 296e64c6-2446-4883-8696-43c72785f01e.jpg
-rw-r--r-- 1 root root 61813 Sep  2 17:04 35130771-dd4a-4646-84a2-326adb21bb57.jpg
-rw-r--r-- 1 root root 92624 Sep  2 17:08 372d6139-792c-4b1b-b565-cdeb734ffb36.jpg
-rw-r--r-- 1 root root 60431 Sep  2 17:04 4dd1b597-8114-4e1e-bd06-507f802bf433.jpg
-rw-r--r-- 1 root root 66416 Sep  2 16:09 5223581e-fff2-4b63-a29a-612308236a8f.jpg
-rw-r--r-- 1 root root 60520 Sep  3 07:25 56c34887-2db4-45c3-aae7-4af0d8e4d0bd.jpg
-rw-r--r-- 1 root root 53066 Sep  2 15:08 570a8b8c-52a1-46b4-ac61-e39ada3b9ab5.jpg
-rw-r--r-- 1 root root 74920 Sep  2 18:10 63c61ccb-85ac-4309-abfd-fd243b67702f.jpg
-rw-r--r-- 1 root root 50653 Sep  2 18:02 772e9dde-918c-42f3-ad70-5a0cc72dd518.jpg
-rw-r--r-- 1 root root 84548 Sep  2 18:21 7b8cb32b-4b90-41b3-9194-08bb7e20ad53.jpg
-rw-r--r-- 1 root root 35699 Sep  2 14:50 8898c0e2-efc1-442c-8bf5-39832b845312.jpg
-rw-r--r-- 1 root root 53623 Sep  3 08:58 b70f134b-f095-4a52-9147-db16e50ff0db.jpg
-rw-r--r-- 1 root root 33573 Sep  2 20:32 da6b9e60-dba2-475e-9c7a-94fb244894eb.jpg
-rw-r--r-- 1 root root 70110 Sep  2 18:10 e0f7a097-ccba-4a4f-b05f-23a5a6d762d0.jpg
-rw-r--r-- 1 root root 55067 Sep  3 08:55 e5320594-dbf6-41ba-a288-d7d60a70d7c9.jpg
-rw-r--r-- 1 root root 73094 Sep  2 17:09 fef65f4b-46d3-4369-ade9-9633ed4876c5.jpg
```

```
# ls -l train/ashley/
total 248
-rw-rw-rw- 1 root root 97861 Sep  2 17:07 007ba212-f92d-4fe4-ba27-b1b3e4bcf230.jpg
-rw-rw-rw- 1 root root 92624 Sep  2 17:08 372d6139-792c-4b1b-b565-cdeb734ffb36.jpg
-rw-rw-rw- 1 root root 60520 Sep  3 07:25 56c34887-2db4-45c3-aae7-4af0d8e4d0bd.jpg
```

```
# ls -l train/manuel/
total 200
-rw-rw-rw- 1 root root 53066 Sep  2 15:08 570a8b8c-52a1-46b4-ac61-e39ada3b9ab5.jpg
-rw-rw-rw- 1 root root 74920 Sep  2 18:10 63c61ccb-85ac-4309-abfd-fd243b67702f.jpg
-rw-rw-rw- 1 root root 70110 Sep  2 18:10 e0f7a097-ccba-4a4f-b05f-23a5a6d762d0.jpg
```

Second function: Thank you very much

**Comment by jakowenko** [comment]
The file name changes when it's copied, just to make sure it doesn't overwrite an existing file in the trianing folder. The file sizes for manuel and ashely seem to match up to file sizes in the matches folder. Are you sure when you go to the Train page in the UI that you don't see the one that was just trained?

**Comment by mr-manuel** [comment]
In my case the files aren't copied. Therefore I copied the files manually for testing. After I train an image from the Matches tab I can't find them under the Train tab.

1) I go to the Train tab -> select ashley on the top left -> select the pictures with ashley on it -> click on train on the top left -> confirm with yes -> after a few seconds the "2 files trained for ashley" message appears on the bottom left. Now I check the train folder for ashley but it's still empty.

2) I manually copy the files to the Train folders in filesystem. After restarting the docker container the images arent't displayed anymore in the Maches tab, even if the files are still in the folder.

**Comment by jakowenko** [comment]
Hm, that's strange. Here's a GIF of how it works on my computer. You can see the train folder is empty and also shows no results on the UI. When I train an image it copies it over then when I go to the train page on the UI the trained image is there as well.

What version of Double Take are you running?

![2021-09-03 14 49 04](https://user-images.githubusercontent.com/1081811/132053343-9acd7c83-c7e0-483a-8934-d727eeccf086.gif)

**Comment by mr-manuel** [comment]
I see. Very strange. Can I enable advanced logging somehow?

I‘m using the latest beta released 6 hours ago.

**Comment by jakowenko** [comment]
I don't have any advanced logging in there at the moment. There is a messages.log file that you could send me on discord if you'd like. That should include all the logs from the system. I can see if anything stands out to me in there.

Very strange though, I haven't heard of this happening to anybody else.

**Comment by jakowenko** [comment]
I should be able to get some verbose logs in early next week. If you aren't able to figure out what's going on over the weekend, hopefully some more logs will help us debug.

**Comment by mr-manuel** [comment]
I think now I got the problem. I'm publishing Double Take with a reverse proxy (Traefik). When I train over HTTP (http://doubletake.local) all works fine. But when I'm training over HTTPS with a internal certificate (https://doubletake.local) it does not work and I get this error in the logs:

```
21-09-04 10:22:20 error: save url error: unable to verify the first certificate
Error: unable to verify the first certificate
    at TLSSocket.onConnectSecure (_tls_wrap.js:1514:34)
    at TLSSocket.emit (events.js:400:28)
    at TLSSocket._finishInit (_tls_wrap.js:936:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:708:12)
```

Maybe we need an option to ignore untrusted certificates?

**Comment by mr-manuel** [comment]
If I put the containers in an isolated network without access to the outside (so no external DNS available, access only through reverse proxy) I get this error:

```
21-09-04 17:38:44 error: save url error: getaddrinfo EAI_AGAIN doubletake.local
Error: getaddrinfo EAI_AGAIN doubletake.local
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:69:26)
21-09-04 17:38:49 error: save url error: getaddrinfo EAI_AGAIN doubletake.local
Error: getaddrinfo EAI_AGAIN doubletake.local
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:69:26)
21-09-04 17:38:54 error: save url error: getaddrinfo EAI_AGAIN doubletake.local
Error: getaddrinfo EAI_AGAIN doubletake.local
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:69:26)
```

This tells me that the saving mechanism is not working internally but an internal process is calling the hostname from the HTTP request. Is this an expected behavior? In my eyes it should use localhost or a host/ip from the config file. What do you think about it?

**Comment by jakowenko** [comment]
Hey @mr-manuel, sorry I was out of town for the weekend.

You should be able to disable the self signed certificate warnings by adding this to your docker-compose.yml file.

```yaml
environment:
    NODE_TLS_REJECT_UNAUTHORIZED: 0
```

The training endpoint takes a full image path, to make it dynamic for images that live outside of Double Take along with images from within. But this seem to cause more problems when trying to train from within the UI itself, especially when reverse proxies and self signed certs are involved. I'll tweak this logic and hopefully that will fix your issue. Rather than passing a full image path, I'll pass just the filename to the API and let the API find and copy the image rather than downloading.

Right now the image path is in this format for you: `https://doubletake.local/api/storage/matches/UUID-.jpg`, so if the container can't resolve this address, then that error would make sense.

**Comment by jakowenko** [comment]
I just pushed a new `beta` build which should fix this for you. The API will copy the image now instead of trying to download it. This should fix the errors you were getting due to the reverse proxy and certs.

Let me know when you test this out if it works.

**Comment by mr-manuel** [comment]
No problem. Thank you very much!

I tested it and it works. It even do not copy an image again, if it was already copied. Before it copied an image over and over again, if I was training the same image multiple times.

In the beta from about 30 minutes ago I found some "HTTP 421 Redirection" issues, but now with the latest beta (released 5 minutes ago) it did not happen again.

Problem solved, thanks again.
